src/main.cpp

```diff
@@ -115,748 +160,765 @@ int main ( int argc, char** argv )
+                    }
 
-                if ( !strServerListFilter.isEmpty() )
-                {
-                    qWarning() << "Server list filter will only take effect when running as a directory server.";
-                    strServerListFileName = "";
+                    if ( !strServerListFilter.isEmpty() )
+                    {
+                        qWarning() << "Server list filter will only take effect when running as a directory server.";
+                        strServerListFileName = "";
+                    }
                 }
-            }
 
-            if ( strDirectoryServer.isEmpty() )
-            {
-                if ( !strServerPublicIP.isEmpty() )
+                if ( strDirectoryServer.isEmpty() )
                 {
-                    qWarning() << "Server Public IP will only take effect when registering a server with a directory server.";
-                    strServerPublicIP = "";
+                    if ( !strServerPublicIP.isEmpty() )
+                    {
+                        qWarning() << "Server Public IP will only take effect when registering a server with a directory server.";
+                        strServerPublicIP = "";
+                    }
                 }
-            }
-            else
-            {
-                if ( !strServerPublicIP.isEmpty() )
+                else
                 {
-                    QHostAddress InetAddr;
-                    if ( !InetAddr.setAddress ( strServerPublicIP ) )
+                    if ( !strServerPublicIP.isEmpty() )
                     {
-                        qWarning() << "Server Public IP is invalid. Only plain IP addresses are supported.";
-                        strServerPublicIP = "";
+                        QHostAddress InetAddr;
+                        if ( !InetAddr.setAddress ( strServerPublicIP ) )
+                        {
+                            qWarning() << "Server Public IP is invalid. Only plain IP addresses are supported.";
+                            strServerPublicIP = "";
+                        }
                     }
                 }
             }
-        }
 
-        if ( !strServerBindIP.isEmpty() )
-        {
-            QHostAddress InetAddr;
-            if ( !InetAddr.setAddress ( strServerBindIP ) )
+            if ( !strServerBindIP.isEmpty() )
             {
-                qWarning() << "Server Bind IP is invalid. Only plain IP addresses are supported.";
-                strServerBindIP = "";
+                QHostAddress InetAddr;
+                if ( !InetAddr.setAddress ( strServerBindIP ) )
+                {
+                    qWarning() << "Server Bind IP is invalid. Only plain IP addresses are supported.";
+                    strServerBindIP = "";
+                }
             }
         }
-    }
 
-    // Application/GUI setup ---------------------------------------------------
-    // Application object
-#ifdef HEADLESS
```

Comment:
This is where you should have `QScopedPointer<QCoreApplication> pApp;` - see my branch. You don't need it to be of different types in the headless/non-headless cases, because `QApplication` is a subclass of `QCoreApplication` and has a virtual destructor.

You can then remove all of the extra conditionals that you've added below because of splitting this into two variables.